src/cswinrt/helpers.h

```diff
@@ -994,4 +994,202 @@ namespace cswinrt
+    }
+
+    template <typename T>
+    auto get_attribute_value(CustomAttribute const& attribute, uint32_t const arg)
+    {
+        return std::get<T>(std::get<ElemSig>(attribute.Value().FixedArgs()[arg].value).value);
+    }
+
+    std::optional<int> get_contract_version(TypeDef const& type)
+    {
+        if (!has_attribute(type, "Windows.Foundation.Metadata", "ContractVersionAttribute"))
+        {
+            return {};
+        }
+        return get_attribute_value<uint32_t>(get_attribute(type, "Windows.Foundation.Metadata"sv, "ContractVersionAttribute"sv), 1);
+    }
+
+    std::optional<int> get_version(TypeDef const& type)
+    {
+        if (!has_attribute(type, "Windows.Foundation.Metadata", "VersionAttribute"))
+        {
+            return {};
+        }
+        return get_attribute_value<uint32_t>(get_attribute(type, "Windows.Foundation.Metadata"sv, "VersionAttribute"sv), 0);
+    }
+
+    bool is_fast_abi_class(TypeDef const& type)
+    {
+        return has_attribute(type, "Windows.Foundation.Metadata"sv, "FastAbiAttribute"sv);
+    }
+
+    std::optional<TypeDef> find_fast_abi_class_type(TypeDef const& iface)
+    {
+        auto exclusiveToAttribute = get_attribute(iface, "Windows.Foundation.Metadata"sv, "ExclusiveToAttribute"sv);
+        if (exclusiveToAttribute)
+        {
+            auto sys_type = get_attribute_value<ElemSig::SystemType>(exclusiveToAttribute, 0);
+            TypeDef exclusiveToClass = iface.get_cache().find_required(sys_type.name);
+            if (!has_attribute(exclusiveToClass, "Windows.Foundation.Metadata"sv, "FastAbiAttribute"sv))
+            {
+                return {};
+            }
+            return exclusiveToClass;
+        }
+        return {};
+    }
+
+    std::pair<TypeDef, std::vector<TypeDef>> get_default_and_exclusive_interfaces(TypeDef const& classType)
+    {
+        std::pair<TypeDef, std::vector<TypeDef>> exclusive_ifaces;
+        std::vector<TypeDef> exclusive_interfaces;
+        for (auto&& ifaceImpl : classType.InterfaceImpl())
+        {
+            auto&& sem = get_type_semantics(ifaceImpl.Interface());
+            if (has_attribute(ifaceImpl, "Windows.Foundation.Metadata", "DefaultAttribute"))
+            {
+                exclusive_ifaces.first = std::get<type_definition>(sem);
+            }
+            else if (std::holds_alternative<type_definition>(sem))
+            {
+                if (has_attribute(std::get<type_definition>(sem), "Windows.Foundation.Metadata"sv, "ExclusiveToAttribute"sv))
+                {
+                    exclusive_ifaces.second.push_back(std::get<type_definition>(sem));
+                }
+            }
+        }
+        return exclusive_ifaces;
+    }
+
+    void sort_fast_abi_ifaces(std::vector<TypeDef>& fast_abi_ifaces)
+    {
+        std::sort(fast_abi_ifaces.begin(), fast_abi_ifaces.end(), [](TypeDef const& iface, TypeDef const& otherIface)
+        {
+            // compare relative contracts
```

Comment:
I don't see the corresponding logic in cppwinrt - no mention of PreviousContractVersion here:
[https://github.com/microsoft/cppwinrt/blob/3393e79d5c405ffee25d468f61f921c6216d21fe/cppwinrt/helpers.h#L451](https://github.com/microsoft/cppwinrt/blob/3393e79d5c405ffee25d468f61f921c6216d21fe/cppwinrt/helpers.h#L451)